pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -659,23 +659,51 @@ func endpointPortFromServicePort(servicePort *v1.ServicePort, portNum int) *v1.E
 	return epp
 }
 
-// overCapacity returns true if there are more addresses in the provided subsets
-// than the maxCapacity.
-func overCapacity(subsets []v1.EndpointSubset) bool {
+// capacityAnnotationSetCorrectly returns false if number of endpoints is greater than maxCapacity or
+// returns true if underCapacity and the annotation is not set.
+func capacityAnnotationSetCorrectly(annotations map[string]string, subsets []v1.EndpointSubset) bool {
 	numEndpoints := 0
 	for _, subset := range subsets {
 		numEndpoints += len(subset.Addresses) + len(subset.NotReadyAddresses)
 	}
-	return numEndpoints > maxCapacity
+	if numEndpoints > maxCapacity {
+		// If subsets are over capacity, they must be truncated so consider
+		// the annotation as not set correctly
+		return false
+	}
+	_, ok := annotations[v1.EndpointsOverCapacity]
+	return !ok
 }
 
-// capacityAnnotationSetCorrectly returns true if overCapacity() is true and the
-// EndpointsOverCapacity annotation is set to "warning" or if overCapacity()
-// is false and the annotation is not set.
-func capacityAnnotationSetCorrectly(annotations map[string]string, subsets []v1.EndpointSubset) bool {
-	val, ok := annotations[v1.EndpointsOverCapacity]
-	if overCapacity(subsets) {
-		return ok && val == "warning"
+// truncateEndpoints ensures that the number of Addresses does not exceed the maxCapacity. If greater
+// than maxCapacity, the addresses will be truncated so that the number of addresses is maxCapacity.
+func truncateEndpoints(endpoints *v1.Endpoints) bool {
+	canBeAdded := maxCapacity
+
+	for i := range endpoints.Subsets {
```

Comment:
I'm not 100% sure this is necessary, but it may be safer to switch this around to focus on even subset distribution. Not sure how common this is, but it's theoretically possible that there are 5 subsets each with 1000 endpoints. Instead of just capturing the first subset, I think it would be better if we had 200 from each subset.

Reply:
The only concern I had about trying to even out the distribution was the efficiency, since this is in the critical path of every update. If we were to implement an even distribution it would loop through the endpoints ~3 times, which is still `O(n)`, so maybe not so bad.